tests/string_test.c

```diff
@@ -0,0 +1,310 @@
+
+    ASSERT_SUCCESS(aws_array_list_get_at(&output, &value, 1));
+    expected = "testa";
+
+    ASSERT_BIN_ARRAYS_EQUALS(expected, strlen(expected), value.ptr, value.len);
+
+    ASSERT_SUCCESS(aws_array_list_get_at(&output, &value, 2));
+    expected = "testb;testc";
+    ASSERT_BIN_ARRAYS_EQUALS(expected, strlen(expected), value.ptr, value.len);
+
+    aws_array_list_clean_up(&output);
+
+    return 0;
+}
+
+AWS_TEST_CASE(test_char_split_with_max_splits, test_char_split_with_max_splits_fn)
+
+static int test_char_split_output_too_small_fn(struct aws_allocator *allocator, void *ctx) {
+    const char str_to_split[] = "testa;testb;testc;";
+
+    struct aws_byte_buf to_split = aws_byte_buf_from_literal(str_to_split);
+
+    struct aws_array_list output;
+    struct aws_byte_buf output_array[3] = {{0}};
+    ASSERT_SUCCESS(aws_array_list_init_static(&output, output_array, 3, sizeof(struct aws_byte_cursor)));
+    ASSERT_ERROR(AWS_ERROR_LIST_EXCEEDS_MAX_SIZE, aws_string_split_on_char(&to_split, ';', &output));
+   
+    return 0;
+}
+
+AWS_TEST_CASE(test_char_split_output_too_small, test_char_split_output_too_small_fn)
+
+static int test_buffer_cat_fn(struct aws_allocator *allocator, void *ctx) {
+    struct aws_byte_buf str1 = aws_byte_buf_from_literal("testa");
+    struct aws_byte_buf str2 = aws_byte_buf_from_literal(";testb");
+    struct aws_byte_buf str3 = aws_byte_buf_from_literal(";testc");
+
+
+    const char expected[] = "testa;testb;testc";
+
+    struct aws_byte_buf destination;
+    ASSERT_SUCCESS(aws_byte_buf_alloc(allocator, &destination, str1.len + str2.len + str3.len));
+    ASSERT_SUCCESS(aws_byte_buf_cat(&destination, 3, &str1, &str2, &str3));
+
+    ASSERT_BIN_ARRAYS_EQUALS(expected, strlen(expected), destination.buffer, destination.len);
+
+    aws_byte_buf_free(allocator, &destination);
+
+    return 0;
+}
+
+AWS_TEST_CASE(test_buffer_cat, test_buffer_cat_fn)
+
+static int test_buffer_cat_dest_too_small_fn(struct aws_allocator *allocator, void *ctx) {
+    struct aws_byte_buf str1 = aws_byte_buf_from_literal("testa");
+    struct aws_byte_buf str2 = aws_byte_buf_from_literal(";testb");
+    struct aws_byte_buf str3 = aws_byte_buf_from_literal(";testc");
+
+    struct aws_byte_buf destination;
+    ASSERT_SUCCESS(aws_byte_buf_alloc(allocator, &destination, str1.len + str2.len));
+    ASSERT_ERROR(AWS_ERROR_DEST_COPY_TOO_SMALL, aws_byte_buf_cat(&destination, 3, &str1, &str2, &str3));
+
+    aws_byte_buf_free(allocator, &destination);
+
+    return 0;
+}
+
+AWS_TEST_CASE(test_buffer_cat_dest_too_small, test_buffer_cat_dest_too_small_fn)
+
+static int test_buffer_cpy_fn(struct aws_allocator *allocator, void *ctx) {
+    struct aws_byte_buf from = aws_byte_buf_from_literal("testa");
+    struct aws_byte_buf destination;
+
+    ASSERT_SUCCESS(aws_byte_buf_alloc(allocator, &destination, from.len));
+    ASSERT_SUCCESS(aws_byte_buf_copy(&destination, &from));
```

Comment:
I'm also wondering if we should have the concept of a buffer with both a _maximum_ size and a _current_ size to make it easier to expose the result size from methods like this.

Reply:
I’m curious what value that would have since it’s always going to be from->len

Reply:
done.